### ISSUE TITLE
`map` over several tuples demands equal length

### ISSUE THREAD

**Issue by mcabbott** [JuliaLang/julia#42216]
Instead of stopping early like `zip`, this is an error:
```julia
julia> map(+, (1,2,3), (4,5))
ERROR: BoundsError: attempt to access Tuple{} at index [1]
Stacktrace:
 [1] getindex(t::Tuple, i::Int64)
   @ Base ./tuple.jl:29
 [2] map (repeats 3 times)
   @ ./tuple.jl:250 [inlined]
 [3] top-level scope
   @ REPL[32]:1

julia> map(+, (1,2,3), [4,5])  # what I expected
2-element Vector{Int64}:
 5
 7
```
Is this a bug? It seems to be present at least since 1.0. The docstring for `map` says

> For multiple collection arguments, apply `f` elementwise, and stop when when any of them is exhausted.

but I'm not sure I can quote it as authoratative, since I think I wrote it.

**Comment by Seelengrab** [comment]
Copying from slack:

If it is, it could be made consistent by adding these:

```
julia> @eval Base map(f, ::Tuple{}, ::Tuple{Any}) = ()
map (generic function with 61 methods)                
                                                      
julia> @eval Base map(f, ::Tuple{Any}, ::Tuple{}) = ()
map (generic function with 62 methods)                
                                                      
julia> map(+, (1,2,3), (4,5))                         
(5, 7)                                                
```

though that won't fix the 3+ argument case. Might just need a `any(isempty, tails) ? () : map(f, tails...)` in the right place for that, unsure how much inference likes this.

**Comment by timholy** [comment]
I love that it throws an error, and I wish `zip` did too.

**Comment by jakobnissen** [comment]
For usability, I think it's great that it throws an error, and it would be nice if `map` did that on arrays as well. Better to be strict by default - and when you're mapping over N containers, it's presumably the intention that every element of every container participates in the mapping. It's much nicer to have map error because you violated an assumption about your data instead of tracking down an error caused by map silently only mapping some of the elements of the collection.

The docstring was changed after the release of 1.6, so haven't appeared in any released Julia versions, so I don't think it's breaking to change it back.

Arguably, it's the correct behaviour to error, since the canonical docstring says "[ ... ] by applying `f` to _each element_". If it only applies to the first elements, it doesn't apply each element.

**Comment by mcabbott** [comment]
For once I actually wanted the early-stopping behaviour on tuples, yesterday. The fact that apparently nobody else did, for years, does make it seem like this isn't widely used. (And it's obviously easy to roll your own.)

Agree that the stricter behaviour would avoid some surprises. But changing the behaviour for arrays (and iterators) sounds like it would be breaking. It just collects `Base.Generator(+, [1,2,3], [4,5])`, which contains `zip`; I'm not sure that the fact that nobody got around to updating the docstring until recently means this can be freely changed. 

For arrays note also that `map(+, ones(2,2), ones(2,1))` is an error, while `map(+, ones(2,2), ones(2,1,1))` is a zip. (Same since 1.0, but also no docstring until recently.)

**Comment by mcabbott** [comment]
The present behaviour for `map` on Tuples is tested to give an error here:
https://github.com/JuliaLang/julia/blob/master/test/tuple.jl#L265

I don't actually see a test for `map` on arrays / iterators stopping early. (Although I have not run all tests.) Would it be crazy to change that to demand they all have the same length, or same size even? That would certainly be simpler to think about.

I doubt that `zip` could change its default, but it could gain an option. Then something like `map(f, iters...) = collect(Generator(splat(f), zip(iters...; strict=true)))` would be a generic `map` which behaves more like the Tuple case does now.

**Comment by stevengj** [comment]
For `zip`, see the discussion at #20499

**Comment by o314** [comment]
> I love that it throws an error, and I wish zip did too.

python, ruby, haskell, rust ... do no error when lengths mismatch on zip call, so i think this is not a good reco.

**Comment by timholy** [comment]
`zip` definitely can't change its default, at least not until Julia 2.0. Gaining a `strict=true` option seems like the way to go here.

**Comment by JeffBezanson** [comment]
I have always been a fan of allowing different lengths. But here's a point from triage: it might make sense to check up front and require length/shape matches for anything that has a definitive shape like an array, but for iterators allow mismatches. The reasoning is that with iterators, you have no choice but to start mapping, and only later find out whether the lengths matched, and it seems silly to "throw away" all the work.
